mace/modules/models.py

```diff
@@ -254,120 +187,364 @@ def forward(
+        compute_virials: bool = False,
+        compute_stress: bool = False,
+        compute_displacement: bool = False,
+    ) -> Dict[str, Optional[torch.Tensor]]:
+        """Forward pass - evaluation of model
+
+        Parameters
+        ----------
+        data
+            input graphs, as dictionary
+        training
+            If we are in training mode. Computational graphs are retained, for reuse.
+        compute_force
+            Compute forces on all nodes
+        compute_virials, compute_stress, compute_displacement
+            compatibility only, RAISES ERROR IF True
+
+        Returns
+        -------
+        results
+            calculated results
+            - `energy`: total energy on each graph
+            - `node_energy`: energy of each node, i.e. local energy on each atom
+            - `forces`: negative gradient of total energy on each atom (node)
+            - `dipole`: dipole of the graph, includes atomic and fixed charge dipole
+                components as well
+            - `atomic_dipoles`: dipoles of each atom
+
+        """
+        # dipoles and virials / stress not supported simultaneously
+        error_msg = "dipoles and virials / stress not supported simultaneously"
+        assert not compute_virials, error_msg
+        assert not compute_stress, error_msg
+        assert not compute_displacement, error_msg
 
-        # Add E_0 and (scaled) interaction energy
-        total_energy = e0 + inter_e
-        node_energy = node_e0 + node_inter_es
+        # Setup
+        data["positions"].requires_grad = True
 
-        forces, virials, stress = get_outputs(
-            energy=inter_e,
+        # Evaluate layer outputs & unpack
+        layer_outputs = self._calculate_layer_interactions(data)
+        interaction_energies = layer_outputs[:, 0]
+        atomic_dipoles = layer_outputs[:, 1:]
+
+        # Calculate energies
+        total_energy, node_energy = self._calc_energy(data, interaction_energies)
+
+        # Calculate derivatives if needed
+        forces, _, _ = get_outputs(
+            energy=total_energy,
             positions=data["positions"],
-            displacement=displacement,
+            displacement=None,
             cell=data["cell"],
             training=training,
             compute_force=compute_force,
             compute_virials=compute_virials,
             compute_stress=compute_stress,
         )
 
-        output = {
+        return {
             "energy": total_energy,
             "node_energy": node_energy,
-            "interaction_energy": inter_e,
             "forces": forces,
-            "virials": virials,
-            "stress": stress,
-            "displacement": displacement,
+            "dipole": self._calc_total_dipole(data, atomic_dipoles),
+            "atomic_dipoles": atomic_dipoles,
         }
 
-        return output
+
+@compile_mode("script")
+class ScaleShiftEnergyDipoleMACE(EnergyDipolesMACE, ScaleShiftEnergyModelMixin):
```

Comment:
this is just an extra, which was easy to add due to the separated logic of dipoles & energy